Fix several race conditions when importing files form the ESP

aplanas opened the pull request.

aplanas committed:
Avoid race condition when multiple disks are encrypted

Signed-off-by: Alberto Planas <aplanas@suse.com>

aplanas committed:
Use a generator to mount ESP

Signed-off-by: Alberto Planas <aplanas@suse.com>

aplanas commented:
cc: @arvidjaar

Vogtinator commented on pcr-signature-generator.sh:
```diff
@@ -0,0 +1,49 @@
+Wants=cryptsetup-pre.target
+Before=cryptsetup-pre.target
+
+Wants=boot-efi.mount
```
`RequiresMountsFor=/boot/efi`?

aplanas replied:
Works. Done.

Vogtinator commented on pcr-signature-generator.sh:
```diff
@@ -0,0 +1,49 @@
+
+[Unit]
+Description=Mount /boot/efi
+After=dev-disk-by\\x2dpartuuid-${ESP_UUID//-/\\x2d}.device
```
Should be implicit

aplanas replied:
I am not sure : ( The generator for boot-esp.mount has this after for a target that represent the device:

```
# Automatically generated by systemd-fstab-generator

[Unit]
Documentation=man:fstab(5) man:systemd-fstab-generator(8)
SourcePath=/etc/fstab
Before=local-fs.target
Requires=systemd-fsck@dev-disk-by\x2duuid-C780\x2dCB73.service
After=systemd-fsck@dev-disk-by\x2duuid-C780\x2dCB73.service
After=blockdev@dev-disk-by\x2duuid-C780\x2dCB73.target

[Mount]
What=/dev/disk/by-uuid/C780-CB73
Where=/boot/efi
Type=vfat
Options=utf8
```

aplanas replied:
Now that I read it the `Options=utf8` can also makes sense

Vogtinator replied:
I just did a test here:

```
> cat /run/systemd/system/test.mount
[Mount]
What=/dev/disk/by-uuid/ECF5-000F
Where=/test
Type=vfat
```

```
> systemctl show -P After /test
"dev-disk-by\\x2duuid-ECF5\\x2d000F.device" system.slice systemd-journald.socket local-fs-pre.target "blockdev@dev-disk-by\\x2duuid-ECF5\\x2d000F.target" -.mount
> systemctl show -P Requires /test
"dev-disk-by\\x2duuid-ECF5\\x2d000F.device" system.slice -.mount
```

aplanas replied:
Dropping then

aplanas replied:
Done

Vogtinator commented on pcr-signature-generator.sh:
```diff
@@ -0,0 +1,49 @@
+ESP_UUID="$(dd "if=$EFIVAR" bs=2 skip=2 conv=lcase status=none | tr -d '\0')"
+DEV="/dev/disk/by-partuuid/${ESP_UUID}"
+
+cat > "$1/boot-efi.mount" <<EOF
```
Most generators use `/tmp` if $1 is not set. Currently it would fail with an undefined variable error

Vogtinator replied:
(Ironically using `/tmp` is actually a security issue - it touches predictable paths in a world-writable directory)

aplanas replied:
Interesting, according to https://www.freedesktop.org/software/systemd/man/latest/systemd.generator.html this first parameter is not optional. I can fail if it is not set, that would make sense

aplanas replied:
Checking presence of $1. Done

Vogtinator commented on pcr-signature.sh:
```diff
@@ -1,35 +1,13 @@
-
-mount -o ro "$DEV" "$MNT" || {
-    echo "pcr-signature: unable to mount ESP"
+# Avoid race condition when multiple disks are encrypted
```
How can this happen if it's called by a service that only runs once?

IMO the check is ok, in case the script is called manually, but I don't see a race otherwise.

aplanas replied:
Ideally this should not happen anymore (single service and ConditionPathExist). I will keep the check but I will drop the comment

Vogtinator commented:
One potential issue: This new `/boot/efi` mount is not stopped in the initrd, only during initrd "shutdown" for switch-root. If something else tries to mount `/boot/efi`, it'll fail, e.g. if you have `x-initrd.mount` set for `/boot/efi` in `/etc/fstab`.

aplanas commented:
> One potential issue: This new `/boot/efi` mount is not stopped in the initrd,

Yes. I was not sure how bad was this. I would expect that other mount services will simply be deactivated, as the device is already mounted.

I still do not know how I should address this. Do you know if some service can be marked to be stopped when initrd is done?

Vogtinator commented:
> > One potential issue: This new `/boot/efi` mount is not stopped in the initrd,
> 
> Yes. I was not sure how bad was this. I would expect that other mount services will simply be deactivated, as the device is already mounted.
> 
> I still do not know how I should address this. Do you know if some service can be marked to be stopped when initrd is done?

That is really tricky, as dependencies otherwise cause the unmount to stop other services, and you end up in a really weird infinitely looping initrd. What you can do is have one unit for mounting and one for unmounting...

arvidjaar commented:
> Do you know if some service can be marked to be stopped when initrd is done?

```
Conflicts=initrd-switch-root.target
Before=initrd-switch-root.target
```
Explicit `Before` makes sense to ensure unit is stopped, otherwise it can race with `Conflicts` unit.

It is also needed for the service itself, to make sure it is marked as stopped before switch to root. I do not think conditions are needed at all - service is expected to run just once and you know that these files do not exist. I would add `RemainAfterExist=true` to make sure it won't be called twice.

If you will go mount unit route (which still sounds like an overkill), it is cleaner to simply ship `pcr-signature.service` in the package. It has no variable parts; it need not be generated. It obscures the implementation for any future contributor. `module-setup.sh` can also install `cryptsetup.wants.d` link. Generator should really only do the steps that are not known when building initrd.

Vogtinator commented:
> Conflicts=initrd-switch-root.target

Wouldn't be enough for the `x-initrd.mount` case. Maybe `Conflicts=initrd-parse-etc.service`?

arvidjaar commented:
> Maybe `Conflicts=initrd-parse-etc.service`?

`initrd-parse-etc.service` is part of the initial transaction, it won't work.

aplanas commented:
> ```
> Conflicts=initrd-switch-root.target
> Before=initrd-switch-root.target
> ```

Added. It is working for this case.

> Wouldn't be enough for the x-initrd.mount case

That is true .. I wonder if a solution is to mount this ESP in a different location then, so the one in the fstab would not collide?

> it is cleaner to simply ship `pcr-signature.service` in the package. It has no variable parts; it need not be generated.

I agree. I extracted it to make it explicit that is static and renamed the generator

> It obscures the implementation for any future contributor. `module-setup.sh` can also install `cryptsetup.wants.d` link. 

Done, but this clearly makes the implementation more cumbersome, as is now hidden in some dracut functions (also ln_r cannot be used, as seems to create the wrong link)

Vogtinator commented:
> > ```
> > Conflicts=initrd-switch-root.target
> > Before=initrd-switch-root.target
> > ```
> 
> Added. It is working for this case.
> 
> > Wouldn't be enough for the x-initrd.mount case
> 
> That is true .. I wonder if a solution is to mount this ESP in a different location then, so the one in the fstab would not collide?

That's already the case - `/boot/efi` vs. `/sysroot/boot/efi`. Filesystems can only be mounted once (exceptions apply...), which is the issue here.

> > it is cleaner to simply ship `pcr-signature.service` in the package. It has no variable parts; it need not be generated.
> 
> I agree. I extracted it to make it explicit that is static and renamed the generator
> 
> > It obscures the implementation for any future contributor. `module-setup.sh` can also install `cryptsetup.wants.d` link.
> 
> Done, but this clearly makes the implementation more cumbersome, as is now hidden in some dracut functions (also ln_r cannot be used, as seems to create the wrong link)

My preference is to add an `[Install]` section and use `$SYSTEMCTL -q --root "$initdir" enable foobar.service`

arvidjaar commented:
> Filesystems can only be mounted once

You can mount device several times read-only. You can mount them only once read-write. For this reason, I think using explicit mount unit complicates things without any clear benefit. Just use `pcr-signature.service` that implicitly mounts and immediately unmounts ESP, order it `Before=local-fs-pre.target` so it will not clash with normal mounts, make this service `RemainAfterExit=true` to ensure it is executed just once.

Having mount unit is better for troubleshooting as it makes workflow obvious. Maybe it can be unmounted in `ExecStopPost` of the `pcr-signature.service`, again, ordering it `Before=local-fs-pre.target` should avoid conflict with normal filesystems.

aplanas commented:
> My preference is to add an `[Install]` section and use `$SYSTEMCTL -q --root "$initdir" enable foobar.service`

Done. It is clearer now.

aplanas commented:
> Having mount unit is better for troubleshooting as it makes workflow obvious.

That was my thinking. Separate the pcr-signature in two services makes the code smaller.

> Maybe it can be unmounted in `ExecStopPost` of the `pcr-signature.service`, again, ordering it `Before=local-fs-pre.target` should avoid conflict with normal filesystems.

You mean `ExecStartPost` maybe? In any case this produces a cycle

aplanas commented:
> That was my thinking. Separate the pcr-signature in two services makes the code smaller.

... and the race condition more clearly fixed, I should add.
